[ABLD-158] Add rules to build openssl from generated configuration (for macos/arm).

aiuto opened the pull request:
Add bazel rules to build openssl from the generated configuration for macos/arm64

```
bazel build @openssl3//:libcrypto
```


Highlights:
- a BUILD file for the library
- module overlay generation
- buildable, but not hooked into CI
- Hand tested that all .o files in omnibus and bazel builds of libcrypto.a are within expected delta of each other.


Next steps:
- Add macos x86
- Add linux arm and x86 configurations
- Add windows

https://datadoghq.atlassian.net/browse/ABLD-158

alopezz commented on bazel/tools/BUILD.bazel:
```diff
@@ -0,0 +1,21 @@
+load("@bazel_skylib//:bzl_library.bzl", "bzl_library")
```
Should these go under bazel/rules instead? Or even the other way around (as what seem to have over there currently seems to be a macro and a target for a tool)? Let's just agree on some convention and stick to it.

aiuto replied:
golden_file_test definitly in rules.
I prefer tools for the things I don't think normal people should use.
I'll move g_f_t and lock down the other.

aiuto committed:
Add initial directory structure to build openssl from generated configuration.

```
bazel build @openssl3//:libcrypto
```

This gets the basic structure but is not good enough to use yet.

Highlights:
- script to generate the http_archive stanza with an files block.
  The files arg can not be loaded from a .bzl file, so it is hard to make
  it look nice. That leads to the solution of making a per-library MODULE.bazel.
- golden file test to make sure the http_archive stanza is up to date.
- include this generated module file from MODULE.bazel

Next steps:
- Add x86 configurations and adjust the select cases.
- Add winows

aiuto committed:
Merge branch 'main' into aiuto/ossl1

aiuto committed:
Merge branch 'main' into aiuto/ossl1

aiuto committed:
Merge branch 'main' into aiuto/ossl1

aiuto committed:
stop building vpsm4-armv8

aiuto committed:
trim

aiuto committed:
revz

aiuto committed:
Merge branch 'main' into aiuto/ossl1

aiuto committed:
Merge branch 'main' into aiuto/ossl1

aiuto committed:
Merge branch 'main' into aiuto/ossl1

aiuto committed:
Merge branch 'main' into aiuto/ossl1

aiuto committed:
Merge branch 'main' into aiuto/ossl1

aiuto committed:
Merge branch 'main' into aiuto/ossl1

aiuto committed:
Merge branch 'main' into aiuto/ossl1

aiuto committed:
Merge branch 'main' into aiuto/ossl1

aiuto committed:
move files

aiuto committed:
move oss to srcs based lists

aiuto committed:
arm/info.c vs. crypto/info.c

aiuto committed:
Makefile

aiuto committed:
Merge branch 'main' into aiuto/ossl1

aiuto committed:
Merge branch 'main' into aiuto/ossl1

aiuto committed:
Merge branch 'main' into aiuto/ossl1

aiuto committed:
Merge branch 'main' into aiuto/ossl1

aiuto committed:
Merge branch 'main' into aiuto/ossl1

aiuto committed:
get ssl build flags right

aiuto committed:
better flags and lib_contents.bzl

aiuto committed:
Merge branch 'main' into aiuto/ossl1

aiuto committed:
dont put the linux stuff in this PR

aiuto committed:
update MODULE for openssl3

aiuto committed:
add missing -DMD5_ASM

aiuto committed:
add unroll8 arm

aiuto committed:
remove alias

aiuto committed:
Merge branch 'main' into aiuto/ossl1

aiuto committed:
add back unroll

aiuto committed:
per arch lib_contents.bzl

aiuto committed:
Merge branch 'main' into aiuto/ossl1

rdesgroppes commented on deps/openssl3/BUILD.bazel:
```diff
@@ -0,0 +1,19 @@
+    url = "https://www.openssl.org/source/openssl-3.5.4.tar.gz",
+)
+
+diff_test(
```
much appreciated!

rdesgroppes reviewed (approved):
shoot!

aiuto commented on deps/openssl3/overlay/overlay.BUILD.bazel:
```diff
@@ -0,0 +1,115 @@
+])
+
+VERSION = "3.5.2"
+SONAME = "libcrypto.so.%s" % VERSION
```
We're going to have to clean this later anyway.

aiuto commented on deps/openssl3/overlay/darwin_arm64/crypto/modes/aes-gcm-armv8-unroll8_64.S:
```diff
@@ -0,0 +1,8487 @@
+	aese	v1.16b, v28.16b
+	aesmc	v1.16b, v1.16b			//AES block 8k+9 - round 5
+	aese	v2.16b, v28.16b
+	aesmc	v2.16b, v2.16b			//AES block 8k+10 - round 5
```
Nope. Just got the right copy.
